include/fast_ber/ber_types/Real.hpp

```diff
@@ -50,4 +63,502 @@ class Real
+        {
+            unsigned int next_byte_shifted = static_cast<unsigned int>(double_bytes[j - 1]) << 8;
+            double_bytes[j] = (next_byte_shifted | double_bytes[j]) >> shift;
+        }
+
+    }
+    else if ((double_bytes[1] & 0xF0u) == 0  && exponent > 0)
+    {
+        /* Mantissa is too small - shift left up to 4 bits */
+        int shift = 8;
+        unsigned int first_byte = double_bytes[1];
+
+        /* Adapt exponent */
+        while ((first_byte & 0xF0) == 0 && exponent > (8 - shift))
+        {
+            exponent -= 8 - shift;
+        }
+
+        for (int j = 1 ; j < 8 ; ++j)
+        {
+            unsigned int first_byte_shifted = static_cast<unsigned int>(double_bytes[j]) << 8;
+            double_bytes[j] = (first_byte_shifted | double_bytes[j + 1]) >> shift;
+        }
+    }
+
+    /* Verify that exponent does not overflow after modification */
+    if (exponent >= 0x7FF || exponent <= -55)
+    {
+        /* Exponent too large for a double */
+        if (exponent >= 0)
+            output = mantissa_sign >= 0 ? std::numeric_limits<double>::infinity() : -std::numeric_limits<double>::infinity();
+            /* Exponent too small for a double */
+        else
+            output = mantissa_sign >= 0 ? 0.0 : -0.0;
+
+        return true;
+    }
+
+    /* Normalize the value if necessary */
+    if (exponent <= 0)
+    {
+        /* correct the exponent to shift one bit more right */
+        --exponent;
+
+        /* Shift the mantissa right until exponent is 0 */
+        int shift = (-exponent) >> 3;
+        for (int j = 7 ; j > shift ; --j)
+        {
+            double_bytes[j] = double_bytes[j - shift];
+        }
+        for (int j = shift ; j > 0; --j)
+        {
+            double_bytes[j] = 0;
+        }
+        /* shift right bit by bit */
+        shift = -exponent & 0x07;
+        for (int j = 7 ; j > 0 ; --j)
+        {
+            unsigned int next_byte_shifted = static_cast<unsigned int>(double_bytes[j - 1]) << 8;
+            double_bytes[j] = (next_byte_shifted | double_bytes[j]) >> shift;
+        }
+
+        /* exponent for a denormalized value */
+        exponent = 0;
+    }
+
+    /* Finally add the mantissa sign and the exponent to the double bytes */
+    double_bytes[0] = (mantissa_sign << 7) | (exponent >> 4);
+    double_bytes[1] = (exponent << 4) | (double_bytes[1] & 0x0F);
+
+    memcpy(&output, double_bytes.data(), sizeof(double));
+    return true;
+}
+
+/**
+ * Decodes a real decimal value
+ * @param input BER encoded real decimal value
+ * @param output decoded decimal value
+ * @return true on success, false on failure
+ */
+inline bool decode_real_decimal(absl::Span<const uint8_t> input, double& output) noexcept
+{
+    std::string iso_6093_value(reinterpret_cast<const char*>(&input[1]), input.length() - 1);
+
+    /* Check that string matches the iso 6093 NR1/NR2/NR3 forms */
+    std::regex iso_6093_nr1 { "^[ ]*[+-]?[0-9]+$" };
```

Comment:
Should be globals variables so that they're not constructed on every decode

Reply:
Maybe we should offer the option to skip these checks alltogether, for performance